src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,779 @@
+  type t = Equality.t Array.t [@@deriving eq, ord]
+
+  let show m =
+    Array.fold_right (fun k result -> Equality.show k ^ "\n" ^ result) m ""
+
+  let hash : t -> int = Array.fold_left (fun acc a -> 31 * acc + Equality.hash a) 0
+
+  let empty () = [||]
+
+  let make_empty_array len = Array.init len (fun i -> (Some i, Z.zero))
+
+  let add_variables_to_domain m indexes = (* add new variables to domain with particular indices; translates old indices to keep consistency *)
+    (* the semantic of indexes can be retrieved from apron: https://antoinemine.github.io/Apron/doc/api/ocaml/Dim.html *)
+    if length indexes = 0 then m else
+      let next_offset_bump_list = (* an ascending list of indices, where the offset is bumped by 1 *)
+        Array.to_list indexes @ [ Array.length m ] (* terminate list with m to avoid out of bounds access *)
+      in
+      let offset_map = Array.make (Array.length m) 0 (* maps each variable to the number of variables that are added before this variable *)
+      in
+      let rec shift (offset, list) index = (* bumps offset & pops list, if/while index is heading the list *)
+        if index = List.hd list then
+          shift (offset+1, List.tl list) index
+        else (offset, list)
+      in
+      let _ =
+        Array.fold_lefti (* iterates over all indices of offset_map, overwrites content with current offset wrt. potential shift *)
+          (fun (offset, offset_bump_list) index _ ->
+             let newoffset, newlist = shift (offset, offset_bump_list) index in
+             offset_map.(index) <- newoffset;
+             (newoffset, newlist))
+          (0, next_offset_bump_list) offset_map
+      in
+      let add_offset_to_array_entry (var, offs) = (* uses offset_map to obtain a new var_index, that is consistent with the new reference indices *)
+        Option.map (fun var_index -> var_index + offset_map.(var_index)) var, offs in
+      let m' = make_empty_array (length m + length indexes)
+      in Array.iteri (fun j eq -> m'.(j + offset_map.(j)) <- add_offset_to_array_entry eq) m;
+      m'(* produces a consistent new conj. of equalities *)
+
+  let remove_variables_from_domain m indexes =
+    let nr_removed_colums = length indexes in
+    if nr_removed_colums = 0 || length m = 0 then m
+    else
+    if length m = nr_removed_colums then [||] else
+      let offset_map = Array.make (Array.length m) 0
+      (* maps each variable to the number of variables that are removed before this variable *)
+      in let _ = Array.fold_lefti
+             (fun offset index _ ->
+                if offset < nr_removed_colums && indexes.(offset) = index then offset + 1
+                else (offset_map.(index) <- offset; offset))
+             0 offset_map in
+      let remove_offset_from_array_entry (var, offs) =
+        Option.map (fun var_index -> var_index - offset_map.(var_index)) var, offs in
+      Array.(copy m |>  filteri (fun i _ -> not @@ Array.mem i indexes) (* filter out removed variables*)
```

Comment:
The `copy` is not needed here, as `filteri` creates a new array.

https://github.com/ocaml-batteries-team/batteries-included/blob/master/src/batArray.ml